src/engine/Engine.cpp

```diff
@@ -2455,3 +2464,74 @@ const Preprocessor *Engine::getPreprocessor()
 {
     return &_preprocessor;
 }
+
+void Engine::minimizeHeuristicCost( const Map<unsigned, double>
+                                    &heuristicCost )
+{
+    _tableau->optimizing();
+
+    _heuristicCost = heuristicCost;
+
+    ENGINE_LOG( "Optimizing w.r.t. the current heuristic cost..." );
+    bool localOptimaReached = false;
+    while ( !localOptimaReached )
+    {
+        DEBUG({
+                ENGINE_LOG
+                    ( Stringf( "Current heuristic cost: %f",
+                               computeHeuristicCost( heuristicCost ) ).ascii() );
+                ASSERT( allVarsWithinBounds() );
+            });
+
+        DEBUG( _tableau->verifyInvariants() );
+
+        mainLoopStatistics();
+        if ( _verbosity > 1 &&  _statistics.getNumMainLoopIterations() %
+             GlobalConfiguration::STATISTICS_PRINTING_FREQUENCY == 0 )
+            _statistics.print();
+
+        // If the basis has become malformed, we need to restore it
+        if ( basisRestorationNeeded() )
+        {
+            if ( _basisRestorationRequired == Engine::STRONG_RESTORATION_NEEDED )
+            {
+                performPrecisionRestoration( PrecisionRestorer::RESTORE_BASICS );
+                _basisRestorationPerformed = Engine::PERFORMED_STRONG_RESTORATION;
+            }
+            else
+            {
+                performPrecisionRestoration( PrecisionRestorer::DO_NOT_RESTORE_BASICS );
+                _basisRestorationPerformed = Engine::PERFORMED_WEAK_RESTORATION;
+            }
+
+            _numVisitedStatesAtPreviousRestoration = _statistics.getNumVisitedTreeStates();
+            _basisRestorationRequired = Engine::RESTORATION_NOT_NEEDED;
+            continue;
+        }
+
+        // Restoration is not required
+        _basisRestorationPerformed = Engine::NO_RESTORATION_PERFORMED;
+
+        // Possible restoration due to preceision degradation
+        if ( shouldCheckDegradation() && highDegradation() )
+        {
+            performPrecisionRestoration( PrecisionRestorer::RESTORE_BASICS );
+            continue;
+        }
+
+        localOptimaReached = performSimplexStep();
+    }
+    _tableau->notOptimizing();
+    ENGINE_LOG( "Optimizing w.r.t. the current heuristic cost - done\n" );
+}
+
+double Engine::computeHeuristicCost( const Map<unsigned, double> &heuristicCost )
```

Comment:
Should we perhaps move this functionality to the cost function manager?

Reply:
Done.